src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -164,15 +167,15 @@ protected async virtual Task<Solution> GetChangedSolutionAsync(CancellationToken
         /// Computes changes for a single document.
         /// Override this method if you want to implement a <see cref="CodeAction"/> subclass that changes a single document.
         /// </summary>
-        protected virtual Task<Document> GetChangedDocumentAsync(CancellationToken cancellationToken)
+        protected virtual Task<Document?> GetChangedDocumentAsync(CancellationToken cancellationToken)
```

Comment:
this doesn't feel like an improvement to me.  i would much rather us say that this should return a `Document`, and then potentially have hte impl actually be resilient to `null` being passed in.  In other words, target the behavior and common case we want people to be coding against, *not* try to exactly be precise with what is ok. 

This makes the common case of all our code (where a Doc *is* returned) nice and simple, and it's only the rarer cases of returning `null` where someone has to do extra work.

Reply:
Note: i feel pretty strongly about this, and would want to discuss this philosophy further with the team before proceeding and making these sorts of changes to public APIs.

Reply:
Is the concern mostly that we have to add all the conversions from `Task<Document>` to `Task<Document?>`? The LDM was discussing that; maybe this is good evidence we need that addressed sooner.

Reply:
it's more that i would actuall ylike to tell people: no, if you implement this, you should return non-null.

i.e. annotate to the state of the world we want (though we can be resilient to code that might have been returning null from before).  I would prefer these APIs just push people toward a much saner pattern of "if you have no change, just return what you were given".

Reply:
> Note: i feel pretty strongly about this, and would want to discuss this philosophy further with the team before proceeding and making these sorts of changes to public APIs.

To be clear, I didn't change the public API. If we make this require a non-null return, then it would be a change to the public API.

Reply:
> If we make this require a non-null return, then it would be a change to the public API. 

I'm specifically stating that the NRT annotation should indicate it's not-null.  The binary-signature (and impl) would still allow and check for null.  

> Other options exist which cause this API to pretend to have one behavior while actually having a different behavior

Yes.  That's a core desire of mine with NRT.  NRT is *not* to describe your behavior, it's to describe your intent to the user of how they should use this thing.  Your behavior can be differnet, and your behavior does need to be resilient to people ignoring your intent.